### PR TITLE
feat(blocks): Add pinecone and jina blocks

### PR DESCRIPTION
**Pull request by aarushik93**
### Background

A vector db, chunking and embeddings are the bread and butter of a lot of AI apps

### Changes 🏗️

Added pincone blocks for vector DB and jina for chunking and embeddings

Only for the new autogpt platform, currently in autogpt_platform/

- Create from scratch and execute an agent with at least 3 blocks
- Import an agent from file upload, and confirm it executes correctly
- Upload agent to marketplace
- Import an agent from marketplace and confirm it executes correctly
- Edit an agent from monitor, and confirm it executes correctly

### CODE REVIEW

File: autogpt_platform/backend/backend/blocks/jina_chunking.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from backend.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from backend.data.model import BlockSecret, SchemaField, SecretField
+
+
+class JinaChunkingBlock(Block):
+    class Input(BlockSchema):
+        texts: list = SchemaField(description="List of texts to chunk")
+        api_key: BlockSecret = SecretField(
+            key="jina_api_key", description="Jina API Key"
+        )
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/jina_chunking.py]: use the api credentials instead

https://docs.agpt.co/server/new_blocks/#blocks-with-authentication

File: autogpt_platform/backend/backend/blocks/pinecone.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+from backend.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from backend.data.model import BlockSecret, SchemaField, SecretField
+
+
+class PineconeInitBlock(Block):
+    class Input(BlockSchema):
+        api_key: BlockSecret = SecretField(
+            key="pinecone_api_key", description="Pinecone API Key"
+        )
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/pinecone.py]: use credential store instead

File: autogpt_platform/backend/backend/blocks/pinecone_query.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+from backend.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from backend.data.model import BlockSecret, SchemaField, SecretField
+
+
+class PineconeQueryBlock(Block):
+    class Input(BlockSchema):
+        api_key: BlockSecret = SecretField(
+            key="pinecone_api_key", description="Pinecone API Key"
+        )
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/pinecone_query.py]: put the two blocks in the same file for jina and pinecone so you have:
- `jina.py`
- `pinecone.py`

and even perhaps a vector db folder around them

**Review by ntindle — CHANGES_REQUESTED** [review]
swap credential methods, and combine the files

File: autogpt_platform/backend/backend/blocks/pinecone_query.py (hunk 3 above)
Comment (aarushik93) [reply autogpt_platform/backend/backend/blocks/pinecone_query.py]: ill group them by use, so: 

vectordb
chunking
embeddings

This makes more sense in this case

File: autogpt_platform/backend/backend/blocks/embeddings/jina.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+from backend.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from backend.data.model import BlockSecret, SchemaField, SecretField, CredentialsMetaInput, CredentialsField
+
+
+class JinaEmbeddingBlock(Block):
+    class Input(BlockSchema):
+        texts: list = SchemaField(description="List of texts to embed")
+        credentials: CredentialsMetaInput[Literal['jina'], Literal['api_key']] = CredentialsField(
+            provider="jina",
+            supported_credential_types={"api_key"}, # noqa
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/embeddings/jina.py]: You shouldn’t need the noqa what’s the issue you’re getting? These type hints are used to decide what loads in the system

File: autogpt_platform/backend/poetry.lock [hunk 5]
```diff
@@ -1892,41 +1994,41 @@ poetry-plugin = ["poetry (>=1.0,<2.0)"]
...
+version = "7.8.0"
+description = "Python Reddit API Wrapper."
 optional = false
-python-versions = "~=3.7"
+python-versions = "~=3.8"
 files = [
-    {file = "praw-7.7.1-py3-none-any.whl", hash = "sha256:9ec5dc943db00c175bc6a53f4e089ce625f3fdfb27305564b616747b767d38ef"},
-    {file = "praw-7.7.1.tar.gz", hash = "sha256:f1d7eef414cafe28080dda12ed09253a095a69933d5c8132eca11d4dc8a070bf"},
+    {file = "praw-7.8.0-py3-none-any.whl", hash = "sha256:66bb2d9dc4f9a35910bf8b4339067c8646ab03270b1c38d77b0487bef8c5c3e2"},
+    {file = "praw-7.8.0.tar.gz", hash = "sha256:910a1478d79ce3892bf35b2c52c90f575d6618dfce248e7bf90eed354c19589c"},
```
Comment (ntindle) [inline autogpt_platform/backend/poetry.lock]: I think this change is breaking? Linter fails on it for me now

File: autogpt_platform/backend/backend/blocks/chunking/jina.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    def __init__(self):
+        super().__init__(
+            id="806fb15e-830f-4796-8692-557d300ff43c",
+            description="Chunks texts using Jina AI's segmentation service",
+            categories={BlockCategory.AI, BlockCategory.TEXT},
+            input_schema=JinaChunkingBlock.Input,
+            output_schema=JinaChunkingBlock.Output,
+        )
+
+    def run(self, input_data: Input, **kwargs) -> BlockOutput:
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/chunking/jina.py]: should be 
```suggestion
    def run(
        self,
        input_data: Input,
        *,
        credentials: APIKeyCredentials,
        **kwargs,
    ) -> BlockOutput:
```

File: autogpt_platform/backend/backend/blocks/chunking/jina.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        url = "https://segment.jina.ai/"
+        headers = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {input_data.credentials.get_secret_value()}",
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/chunking/jina.py]:
```suggestion
            "Authorization": f"Bearer {credentials.api_key.get_secret_value()}",
```

File: autogpt_platform/backend/backend/blocks/chunking/jina.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+from backend.data.model import BlockSecret, SchemaField, SecretField, CredentialsMetaInput, CredentialsField
+
+
+class JinaChunkingBlock(Block):
+    class Input(BlockSchema):
+        texts: list = SchemaField(description="List of texts to chunk")
+
+        credentials: CredentialsMetaInput[Literal['jina'], Literal['api_key']] = CredentialsField(
+            provider="jina",
+            supported_credential_types={"api_key"}, # noqa
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/chunking/jina.py]:
```suggestion
            supported_credential_types={"api_key"},
```

File: autogpt_platform/backend/backend/blocks/embeddings/jina.py (hunk 4 above)
Comment (ntindle) [reply autogpt_platform/backend/backend/blocks/embeddings/jina.py]:
```suggestion
            supported_credential_types={"api_key"},
```

File: autogpt_platform/backend/backend/blocks/embeddings/jina.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            categories={BlockCategory.AI},
+            input_schema=JinaEmbeddingBlock.Input,
+            output_schema=JinaEmbeddingBlock.Output,
+        )
+
+    def run(self,
+        input_data: Input,
+        *,
+        credentials: Credentials,
+        **kwargs) -> BlockOutput:
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/embeddings/jina.py]:
```suggestion
    def run(
        self,
        input_data: Input,
        *,
        credentials: APIKeyCredentials,
        **kwargs,
    ) -> BlockOutput:
    ```

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+from backend.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from backend.data.model import BlockSecret, SchemaField, SecretField, CredentialsMetaInput, CredentialsField
+
+
+class PineconeInitBlock(Block):
+    class Input(BlockSchema):
+        credentials: CredentialsMetaInput[Literal['pinecone'], Literal['api_key']] = CredentialsField(
+            provider="pinecone",
+            supported_credential_types={"api_key"}, # noqa
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]:
```suggestion
            supported_credential_types={"api_key"},
```

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            description="Initializes a Pinecone index",
+            categories={BlockCategory.LOGIC},
+            input_schema=PineconeInitBlock.Input,
+            output_schema=PineconeInitBlock.Output,
+        )
+
+    def run(self, input_data: Input, **kwargs) -> BlockOutput:
+        pc = Pinecone(api_key=input_data.credentials.api_key)
+
+        try:
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]: so on and so forth

**Review by ntindle — CHANGES_REQUESTED** [review]

File: autogpt_platform/backend/backend/blocks/pinecone_query.py (hunk 3 above)
Comment (ntindle) [reply autogpt_platform/backend/backend/blocks/pinecone_query.py]: I understand where you're coming from, but grouping by provider is the typical use so far

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py (hunk 10 above)
Comment (aarushik93) [reply autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]: does it not give your warnings without it?

File: autogpt_platform/backend/backend/blocks/jina/_auth.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def JinaCredentialsField() -> JinaCredentialsInput:
+    """
+    Creates a Jina credentials input on a block.
+
+    """
+    return CredentialsField(
+        provider="jina",
+        supported_credential_types={"api_key"},  # noqa
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/jina/_auth.py]: drop the noqa, shouldn't be needed

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+from backend.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from backend.data.model import CredentialsField, CredentialsMetaInput, SchemaField
+
+
+class PineconeInitBlock(Block):
+    class Input(BlockSchema):
+        credentials: CredentialsMetaInput[Literal["pinecone"], Literal["api_key"]] = (
+            CredentialsField(
+                provider="pinecone",
+                supported_credential_types={"api_key"},  # noqa
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]: drop the noqa

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+from pinecone import Pinecone, ServerlessSpec
+
+from backend.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from backend.data.model import CredentialsField, CredentialsMetaInput, SchemaField
+
+
+class PineconeInitBlock(Block):
+    class Input(BlockSchema):
+        credentials: CredentialsMetaInput[Literal["pinecone"], Literal["api_key"]] = (
+            CredentialsField(
```
Comment (ntindle) [inline autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]: nit: you can do the same thing you did with jina's _auth here but in file if you want. Not needed but just deduplicates stuff

File: autogpt_platform/backend/pyproject.toml [hunk 15]
```diff
@@ -46,6 +46,7 @@ youtube-transcript-api = "^0.6.2"
 googlemaps = "^4.10.0"
 replicate = "^0.34.1"
```
Comment (ntindle) [inline autogpt_platform/backend/pyproject.toml]: line skipped?

**Review by ntindle — CHANGES_REQUESTED** [review]

File: autogpt_platform/backend/backend/blocks/jina/_auth.py (hunk 12 above)
Comment (aarushik93) [reply autogpt_platform/backend/backend/blocks/jina/_auth.py]: it's not breaking but it does give me warnings without it

Comment (ntindle) [reply autogpt_platform/backend/backend/blocks/jina/_auth.py]: it shouldnt!! what warnings

Comment (aarushik93) [reply autogpt_platform/backend/backend/blocks/jina/_auth.py]: Screenshot 2024-10-23 at 7 29 19 PM

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py (hunk 13 above)
Comment (aarushik93) [reply autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]: removed

File: autogpt_platform/backend/backend/blocks/jina/_auth.py (hunk 12 above)
Comment (aarushik93) [reply autogpt_platform/backend/backend/blocks/jina/_auth.py]: removed

File: autogpt_platform/backend/pyproject.toml (hunk 15 above)
Comment (aarushik93) [reply autogpt_platform/backend/pyproject.toml]: removed

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py (hunk 14 above)
Comment (aarushik93) [reply autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]: updated

**Comment by aarushik93** [comment]
I'm not sure how to sort out the lockfile issue. I can't add any new libraries without deleting it and it breaks reddit with the version bump - if anyone has a suggestion lmk pls

File: autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py (hunk 10 above)
Comment (ntindle) [reply autogpt_platform/backend/backend/blocks/vector_dbs/pinecone.py]: not at all

**Comment by ntindle** [comment]
I have a change incomming to the frontend providers in my PR that I think will make the frontends more clear. LMK what you think

**Comment by ntindle** [comment]
Can also pull the jina extract blcok from search.py into here

**Review by ntindle — APPROVED** [review]

**Pull request merged**